### PR TITLE
sql: Extend GRANT/REVOKE for all objects syntax

### PR DESCRIPTION
**Pull request by jkosh44**
This commit extends the syntax for GRANT and REVOKE to allow users to:

  - Grant or revoke privileges on all objects of a certain type within one or more databases.
  - Grant or revoke privileges on all objects of a certain type.

This syntax is not included in PostgreSQL, however it is a useful
extension to GRANT and REVOKE.

### Motivation
This PR adds a feature that has not yet been specified.

### Tips for Reviewer
  * Roughly 1000 of the lines added are tests.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add the syntax to GRANT and REVOKE that allows users to grant or revoke privileges on all objects of a certain type within a database.
  - Add the syntax to GRANT and REVOKE that allows users to grant or revoke privileges on all objects of a certain type.

### CODE REVIEW

File: test/sqllogictest/privilege_grants.slt
Comment (jkosh44) [inline test/sqllogictest/privilege_grants.slt:1]: Locally this file takes me 18 minutes to run.

Comment (mjibson) [reply test/sqllogictest/privilege_grants.slt:1]: Even for a 3k line file that seems too long. Can you track down some of the major slowness?

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -6136,6 +6104,54 @@ impl<'a> Parser<'a> {
...
+                            self.expected(self.peek_pos(), DATABASE, self.peek_token())?;
+                        }
+                        GrantTargetSpecificationInner::AllSchemas {
+                            schemas: self.parse_comma_separated(Parser::parse_schema_name)?,
+                        }
+                    }
+                    _ => unreachable!(),
+                }
+            } else {
+                GrantTargetSpecificationInner::All
```
Comment (mjibson) [inline src/sql-parser/src/parser.rs:6137]: Kinda worried about this one because it's so easily destructive. Does postgres have it?

**Review by mjibson — APPROVED** [review]

File: test/sqllogictest/privilege_grants.slt
Comment (jkosh44) [reply test/sqllogictest/privilege_grants.slt:1]: I'll look into it. Though, I should add that in CI it takes 40 seconds total. My best initial guess is that there's some `debug_assert!` or similar debug only path that is very slow.

Comment (jkosh44) [reply test/sqllogictest/privilege_grants.slt:1]: I spent some time tracking this down, this is almost certainly due to the `has_X_privilege` functions. On a debug build they take approximately 1-3 seconds each, but on a release build they take approximately 0.2 seconds each.

**Pull request merged**